man/seq24.1
.TH Seq24 1 "November 20 2010" "Version 0.9.2" "Seq24 Manual Page"

.SH NAME
Seq24 - Real time MIDI sequencer


.SH SYNOPSIS
.B seq24
[\fIOPTIONS\fP] [\fIFILENAME\fP]


.SH DESCRIPTION
.PP
\fISeq24\fP is a real-time midi sequencer. It was created to
provide a very simple interface for editing and playing
midi 'loops'.


.SH OPTIONS
Seq24 accepts the following options:
.TP 8
.B  \-\-help
Display a list of all commandline options
.TP 8
.B  \-\-version
Display the program version
.TP 8
.B \-\-file \fI<filename>\fP
Load MIDI file on startup
.TP 8
.B \-\-manual_alsa_ports
Seq24 won't attach alsa ports
.TP 8
.B \-\-showmidi
Dumps incoming midi to screen
.TP 8
.B \-\-priority
Runs higher priority with FIFO scheduler
.TP 8
.B \-\-pass_sysex
Passes any incoming sysex messages to all outputs
.TP 8
.B \-\-ignore \fI<number>\fP
Ignore ALSA device <number>
.TP 8
.B \-\-show_keys
Prints pressed key value
.TP 8
.B \-\-interaction_method \fI<number>\fP
Select mouse interaction method
	(0 = seq24) (default)
	(1 = fruity loops method)
.TP 8
.B \-\-jack_transport
seq24 will sync to JACK transport
.TP 8
.B \-\-jack_master
seq24 will try to be JACK master
.TP 8
.B \-\-jack_master_cond
JACK master will fail if there is already a master
.TP 8
.B \-\-jack_start_mode \fI<x>\fP
When seq24 is synced to JACK, the following play modes are available:
	(0 = live mode)
	(1 = song mode) (default)
.TP 8
.B \-\-stats
Print statistics on commandline while running
.TP 8
.B \-\-jack_session_uuid \fI<uuid>\fP
Set uuid for jack session


.SH FILES
\fB$HOME\fP/.seq24rc stores the user settings for seq24.


.SH SUGGESTIONS AND BUG REPORTS
Any bugs found should be reported to the upstream author and/or package 
maintainer.


.SH OTHER INFO
The old project homepage is at http://www.filter24.org/seq24/ the new
one is at https://edge.launchpad.net/seq24/.

It is released under the GNU GPL license.


.SH AUTHOR
Seq24 was written by Rob C. Buse <seq24@filter24.org> and the Seq24team.

This manual page was written by Dana Olson <seq24@ubuntustudio.com> with
additions from Guido Scholz <guido.scholz@bayernline.de>. 


NEWS
seq24-0.9.3 (2016-01-24)

Fixed Bugs
  * Fix LASH support (crash on 64 Bit systems)
  * Fix broken JACK transport with newer jackd version
  * Fix clock tick drift
  * Fix jack session commandline (obsolete --file option removed)

New Features
  * Non recursive make

General Changes
  * Some code cleanups
  * C++11 compatible compiler required


seq24-0.9.2 (2010-11-27)

Fixed Bugs
  * Fix tooltip usage for older GTK versions (GTK_MINOR_VERSION < 12)
  * Fix sched_param memory leaks
  * Fix doubled key event for screen set name line

New Features
  * Add support for jack session, patch provided by Torben Hohn
  * Add interrupt handler for SIGUSR1 to enable LADISH level 1 support
  * Add interrupt handler for SIGINT to ask for unsaved file changes
  * Remove "-f" command line option to be replaced by a simple <filename>
    argument (see "seq24 --help" for more information)

General Changes
  * Add mnemonics for bottom line widgets in main window and label for
    screen set name edit line
  * Add missing command line parameters to help message, display short
    options as well
  * Add command line option for program version
  * Add missing command line parameters to man page
  * Remove complaints about file read error if configuration files do not
    exist
  * Cleanup configure.in: remove unused variables, harmonize option
    enabling/disabling
  * Some code cleanups


seq24-0.9.1 (2010-11-08)

Fixed Bugs
  * Fixed error if path for last used directory is not properly set
  * Fixed adding notes on the beginning when MIDI record is on and not
    playing sequence
  * Fixed buffer overflow caused by string handling (tool menu)
  * Fixed file save as crash
  * Fixed MIDI sysex event reading
  * Fixed error if path for last used directory is not properly set
  * "Cancel" option added to "Save file?" question

New Features
  * Support for MIDI clock synchronization 
  * Support for mute groups
  * Optionally show shortcut key label on sequence icon
  * Fruity loop mouse interaction mode added
  * Monophonic step editing on sequence
  * Polyphonic step editing

General Changes
  * A lot of code cleanups
  * Support for (lost) Win32 platform added
  * Optimized redraw for sequencer roll background (from Win32 version,
    Rob Buse)
  * Adaptations for jack2 applied


seq24-0.9.0 (2008-11-27)

Fixed Bugs
  * Fixed BPM saving on 64 bit platforms.
  * Several string buffer overruns fixed.
  * Several compiler warnings fixed.
  * Fixed memory leak if file loading fails due to format errors.

New Features
  * Round robbin logic on sequence edit added.
  * Start stop shortcut on sequence editor added.
  * Remember last used directory on configuration so you go where
    you worked.
  * Live Quantize, so notes get to the right time when you play.
  * Fixed volume. Makes midi input have a fixed volume when you record.
  * Added Redo function on sequence editor.
  * Switched to sigc++-2.0 API, to make program compatible to latest
    sigc++ version (2.2).
  * Man page from Ubuntu added.
  * Keyboard shortcuts for menu items added.
  * Added backspace as an optional delete key to the pattern editor.
  * Use new file dialog layout, MIDI files are preselectable. ".midi"
    suffix is added if user does not append a valid MIDI file extension.
  * New menu item to show and hide song edit window.
  * Tooltips for bottom line elements in main window added.
  * File new/open/save/close logic rewritten to monitor user applied file
    changes.
  * "dump" program removed.
  * About dialog changed to use new layout.
  * Zooming and panning in the editor window using the scroll wheel (when
    CTRL or SHIFT are used) implemented. Horizontal scroll step interval
    increased to 1/16 note per zoom level and page interval to 1 bar.
  * Panning in the song editor using the scroll wheel (when SHIFT is used)
    implemented.
  * Window icons for song editor and sequence editor added.


src/seq24.cpp
//----------------------------------------------------------------------------
//
//  This file is part of seq24.
//
//  seq24 is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  seq24 is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with seq24; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
//-----------------------------------------------------------------------------

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>

#ifdef __WIN32__
#    include "configwin32.h"
#else
#    include "config.h"
#endif

#include "font.h"
#ifdef LASH_SUPPORT
#    include "lash.h"
#endif
#include "mainwnd.h"
#include "midifile.h"
#include "optionsfile.h"
#include "perform.h"
#include "userfile.h"

/* struct for command parsing */
static struct
option long_options[] = {

    {"help", 0, 0, 'h'},
    {"showmidi", 0, 0, 's'},
    {"show_keys", 0, 0, 'k'},
    {"stats", 0, 0, 'S'},
    {"priority", 0, 0, 'p'},
    {"ignore", required_argument, 0, 'i'},
    {"interaction_method", required_argument, 0, 'x'},
    {"jack_transport",0, 0, 'j'},
    {"jack_master",0, 0, 'J'},
    {"jack_master_cond", 0, 0, 'C'},
    {"jack_start_mode", required_argument, 0, 'M'},
    {"jack_session_uuid", required_argument, 0, 'U'},
    {"manual_alsa_ports", 0, 0, 'm'},
    {"pass_sysex", 0, 0, 'P'},
    {"version", 0, 0, 'V'},
    {0, 0, 0, 0}

};

static const char versiontext[] = PACKAGE " " VERSION "\n";

bool global_manual_alsa_ports = false;
bool global_showmidi = false;
bool global_priority = false;
bool global_device_ignore = false;
int global_device_ignore_num = 0;
bool global_stats = false;
bool global_pass_sysex = false;
Glib::ustring global_filename = "";
Glib::ustring last_used_dir ="/";
std::string config_filename = ".seq24rc";
std::string user_filename = ".seq24usr";
bool global_print_keys = false;
interaction_method_e global_interactionmethod = e_seq24_interaction;

bool global_with_jack_transport = false;
bool global_with_jack_master = false;
bool global_with_jack_master_cond = false;
bool global_jack_start_mode = true;
Glib::ustring global_jack_session_uuid = "";

user_midi_bus_definition   global_user_midi_bus_definitions[c_maxBuses];
user_instrument_definition global_user_instrument_definitions[c_max_instruments];

font *p_font_renderer;

#ifdef LASH_SUPPORT
lash *lash_driver = NULL;
#endif

#ifdef __WIN32__
#   define HOME "HOMEPATH"
#   define SLASH "\\"
#else
#   define HOME "HOME"
#   define SLASH "/"
#endif


int
main (int argc, char *argv[])
{
    /* Scan the argument vector and strip off all parameters known to
     * GTK+. */
    Gtk::Main kit(argc, argv);

    /*prepare global MIDI definitions*/
    for ( int i=0; i<c_maxBuses; i++ )
    {
        for ( int j=0; j<16; j++ )
            global_user_midi_bus_definitions[i].instrument[j] = -1;
    }

    for ( int i=0; i<c_max_instruments; i++ )
    {
        for ( int j=0; j<128; j++ )
            global_user_instrument_definitions[i].controllers_active[j] = false;
    }


    /* Init the lash driver (strip lash specific command line
     * arguments and connect to daemon) */
#ifdef LASH_SUPPORT
    lash_driver = new lash(&argc, &argv);
#endif

    /* the main performance object */
    /* lash must be initialized here because mastermidibus uses the global
     * lash_driver variable*/
    perform p;

    /* read user preferences files */
    if ( getenv( HOME ) != NULL )
    {
        Glib::ustring home( getenv( HOME ));
        last_used_dir = home;
        Glib::ustring total_file = home + SLASH + config_filename;
        
        if (Glib::file_test(total_file, Glib::FILE_TEST_EXISTS))
        {
            printf( "Reading [%s]\n", total_file.c_str());

            optionsfile options( total_file );

            if ( !options.parse( &p ) ){
                printf( "Error Reading [%s]\n", total_file.c_str());
            }
        }

        total_file = home + SLASH + user_filename;
        if (Glib::file_test(total_file, Glib::FILE_TEST_EXISTS))
        {
            printf( "Reading [%s]\n", total_file.c_str());

            userfile user( total_file );

            if ( !user.parse( &p ) ){
                printf( "Error Reading [%s]\n", total_file.c_str());
            }
        }

    }
    else
        printf( "Error calling getenv( \"%s\" )\n", HOME );


    /* parse parameters */
    int c;

    while (true) {

        /* getopt_long stores the option index here. */
        int option_index = 0;

        c = getopt_long(argc, argv, "C:hi:jJmM:pPsSU:Vx:", long_options,
                &option_index);

        /* Detect the end of the options. */
        if (c == -1)
            break;

        switch (c){

            case '?':
            case 'h':

                printf( "Usage: seq24 [OPTIONS] [FILENAME]\n\n" );
                printf( "Options:\n" );
                printf( "   -h, --help: show this message\n" );
                printf( "   -v, --version: show program version information\n" );
                printf( "   -m, --manual_alsa_ports: seq24 won't attach alsa ports\n" );
                printf( "   -s, --showmidi: dumps incoming midi events to screen\n" );
                printf( "   -p, --priority: runs higher priority with FIFO scheduler (must be root)\n" );
                printf( "   -P, --pass_sysex: passes any incoming sysex messages to all outputs \n" );
                printf( "   -i, --ignore <number>: ignore ALSA device\n" );
                printf( "   -k, --show_keys: prints pressed key value\n" );
                printf( "   -x, --interaction_method <number>: see .seq24rc for methods to use\n" );
                printf( "   -j, --jack_transport: seq24 will sync to jack transport\n" );
                printf( "   -J, --jack_master: seq24 will try to be jack master\n" );
                printf( "   -C, --jack_master_cond: jack master will fail if there is already a master\n" );
                printf( "   -M, --jack_start_mode <mode>: when seq24 is synced to jack, the following play\n" );
                printf( "                          modes are available (0 = live mode)\n");
                printf( "                                              (1 = song mode) (default)\n" );
                printf( "   -S, --stats: show statistics\n" );
                printf( "   -U, --jack_session_uuid <uuid>: set uuid for jack session\n" );
                printf( "\n\n\n" );

                return EXIT_SUCCESS;
                break;

            case 'S':
                global_stats = true;
                break;

            case 's':
                global_showmidi = true;
                break;

            case 'p':
                global_priority = true;
                break;

            case 'P':
                global_pass_sysex = true;
                break;

            case 'k':
                global_print_keys = true;
                break;

            case 'j':
                global_with_jack_transport = true;
                break;

            case 'J':
                global_with_jack_master = true;
                break;

            case 'C':
                global_with_jack_master_cond = true;
                break;

            case 'M':
                if (atoi( optarg ) > 0) {
                    global_jack_start_mode = true;
                }
                else {
                    global_jack_start_mode = false;
                }
                break;

            case 'm':
                global_manual_alsa_ports = true;
                break;

            case 'i':
                /* ignore alsa device */
                global_device_ignore = true;
                global_device_ignore_num = atoi( optarg );
                break;

            case 'V':
                printf("%s", versiontext);
                return EXIT_SUCCESS;
                break;

            case 'U':
                global_jack_session_uuid = Glib::ustring(optarg);
                break;

            case 'x':
                global_interactionmethod = (interaction_method_e)atoi(optarg);
                break;

            default:
                break;
        }

    } /* end while */


    p.init();
    p.launch_input_thread();
    p.launch_output_thread();
    p.init_jack();


    p_font_renderer = new font();

    mainwnd seq24_window( &p );
    if (optind < argc)
    {
        if (Glib::file_test(argv[optind], Glib::FILE_TEST_EXISTS))
            seq24_window.open_file(argv[optind]);
        else
            printf("File not found: %s\n", argv[optind]);
    }

    /* connect to lash daemon and poll events*/
#ifdef LASH_SUPPORT
    lash_driver->start(&p);
#endif
    kit.run(seq24_window);

    p.deinit_jack();

    if ( getenv( HOME ) != NULL )
    {
        string home( getenv( HOME ));
        Glib::ustring total_file = home + SLASH + config_filename;
        printf( "Writing [%s]\n", total_file.c_str());

        optionsfile options( total_file );

        if (!options.write( &p))
            printf( "Error writing [%s]\n", total_file.c_str());
    }
    else
    {
        printf( "Error calling getenv( \"%s\" )\n", HOME );
    }

#ifdef LASH_SUPPORT
    delete lash_driver;
#endif

    return EXIT_SUCCESS;
}

